File: src/performance.jl
# Functions to analyze the performance

function accuracy(active_constr_pred::Vector{Vector{Int64}},
                  active_constr_test::Vector{Vector{Int64}})
    n_total = length(active_constr_pred)
    idx_correct = zeros(Int, n_total)
    for i = 1:n_total
        if active_constr_pred[i] == active_constr_test[i]
            idx_correct[i] = 1
        end
    end

    return sum(idx_correct) / n_total, idx_correct
end


function eval_performance(theta::DataFrame,
                          lnr::OT.OptimalTreeClassifier,
                          problem::OptimizationProblem,
                          enc2active_constr::Vector{Vector{Int64}};
                          k::Int64=1   # k best predicted values
                         )

    println("Performance evaluation")
    println("Compute active constraints over test set")

    # Get active_constr for each point
    x_test, y_test, time_test, active_constr_test = solve(theta, problem)

    # Get predicted active constraints for all the test points
    x_pred, y_pred, time_pred, active_constr_pred = predict_best_full(theta, k, lnr, problem, enc2active_constr)

    # Get statistics
    num_var = length(problem.data.c)
    num_constr = length(problem.data.l)
    num_test = size(theta, 1)
    num_train = lnr.prb_.data.features.n_samples
    n_theta = size(theta, 2)
    n_active_sets = length(enc2active_constr)

    # Compute infeasibility and suboptimality
    infeas = [infeasibility(x, problem) for x in x_pred]
    subopt = [suboptimality(x_pred[i], x_test[i], problem) for i = 1:num_test]
    time_comp = [(1 - time_pred[i]/time_test[i])*100 for i = 1:num_test]

    # accuracy
    test_accuracy, idx_correct = accuracy(active_constr_pred, active_constr_test)

    # Get problem name
    if isdefined(problem, :file_name)
        problem_name = lowercase(split(splitdir(problem.file_name)[end], ".")[end-1])
    else
        problem_name = lowercase(split(string(typeof(problem)), ".")[end])
    end

    # DEBUG
    @show sum(infeas .>= TOL)

    # Create dataframe and export it
    df = DataFrame(
                   problem = [problem_name],
                   radius = [problem.radius],
                   k = [k],
                   num_var = Int[num_var],
                   num_constr = Int[num_constr],
                   num_test = Int[num_test],
                   num_train = Int[num_train],
                   n_theta = Int[n_theta],
                   n_correct = Int[sum(idx_correct)],
                   n_active_sets = Int[n_active_sets],
                   accuracy = [test_accuracy],
                   n_infeas = [sum(infeas .>= TOL)],
                   avg_infeas = [mean(infeas)],
                   avg_subopt = [mean(subopt[find(infeas .<= TOL)])],  # Mean of feasible cases
                   max_infeas = [maximum(infeas)],
                   max_subopt = [maximum(subopt[find(infeas .<= TOL)])],
                   avg_time_improvement_perc = [mean(time_comp)],
                   max_time_improvement_perc = [maximum(time_comp)],
                  )

    df_detail = DataFrame(
                          problem = repmat([problem_name], num_test),
                          correct = idx_correct,
                          infeas = infeas,
                          subopt = subopt,
                          time_improvement_perc = time_comp
                         )

    return df, df_detail
end

"""
    infeasibility(x_ml, problem)

Compute infeasibility weighted over the magnitude of l and u.
"""
function infeasibility(x::Vector{Float64},
                       problem::OptimizationProblem)
    l, A, u = problem.data.l, problem.data.A, problem.data.u


    norm_A = [norm(A[i, :], Inf) for i = 1:size(A, 1)]

    upper = max.(A * x - u, 0.)
    lower = max.(l - A * x, 0.)

    # For the non zero ones, normalize
    [upper[i] /= max.(norm_A[i], abs.(u[i])) for i = 1:length(upper) if upper[i] >= TOL]
    [lower[i] /= max.(norm_A[i], abs.(l[i])) for i = 1:length(lower) if lower[i] >= TOL]

    return norm(upper + lower)

end

"""
    suboptimality(x, x_opt, problem)

Compute suboptimality weighted over the magnitude of c.
"""
function suboptimality(x::Vector{Float64},
                       x_opt::Vector{Float64},
                       problem::OptimizationProblem)
    c = problem.data.c
    return (c' * x - c' * x_opt) / norm(c, Inf)
end

function write_output(df::DataFrame,
                      df_detail::DataFrame,
                      problem::OptimizationProblem;
                     output_folder::String="output")
    # Get problem name
    problem_name = lowercase(split(string(typeof(problem)), ".")[end])
    output_name = joinpath(output_folder,
                           problem_name)
    CSV.write("$(output_name).csv", df)
    CSV.write("$(output_name)_detail.csv", df_detail)
    nothing
end

function write_output(df::DataFrame,
                      df_detail::DataFrame;
                      file_name::String="results",
                      output_folder::String="output")
    output_name = joinpath(output_folder, file_name)
    CSV.write("$(output_name).csv", df)
    CSV.write("$(output_name)_detail.csv", df_detail)
    nothing
end

File: mlo/performance.py
import numpy as np
import pandas as pd
from .constants import TOL


def accuracy(strategy_pred,
             strategy_test):
    """
    Accuracy comparison between predicted and test strategies

    Args:
        strategy_pred (Strategy list): List of predicted strategies.
        strategy_pred (Strategy list): List of test strategies.

    Returns:
        float: Fraction of correct over total strategies compared.
        numpy array: Boolean vector indicating which strategy is correct.

    """
    assert len(strategy_pred) == len(strategy_test)
    n_total = len(strategy_pred)
    idx_correct = np.zeros(n_total)
    for i in range(n_total):
        if strategy_pred[i] == strategy_test[i]:
            idx_correct[i] = 1

    return np.sum(idx_correct) / n_total, idx_correct


def eval_performance(theta, lnr, problem, enc2strategy, k=1):
    """
    Evaluate predictor performance

    Parameters
    ----------
    theta : DataFrame
        Data to predict.
    lnr : Learner
        Learner.
    problem : OptimizationProblem
        Optimization problem.
    enc2strategy : Strategy list
        Mapping between encoding to the strategy.
    k : int, optional
        Best k predicted values to pick. Defaults to 0.
    """

    print("Performance evaluation")
    print("Compute active constraints over test set")

    # Get strategy for each point
    x_test, time_test, strategy_test = problem.solve(theta)

    # Get predicted strategy for each point
    x_pred, time_pred, strategy_pred = predict_best(theta, k, lnr,
                                                    problem, enc2strategy)

    num_var = len(problem.data.c)
    num_constr = len(problem.data.l)
    num_test = len(theta)
    num_train = lnr.n_train    # Number of training samples from learner
    n_theta = theta.shape[1]   # Parameters dimension
    n_active_sets = len(enc2strategy)  # Number of active sets

    # Compute infeasibility and optimality for each problem
    # NB. We need to populate the problem for each point
    infeas = []
    subopt = []
    time_comp = []
    for i in range(num_test):
        problem.populate(theta.iloc[i, :])
        infeas.append(problem.infeasibility(x_pred[i]))
        subopt.append(problem.suboptimality(x_pred[i], x_test[i]))
        time_comp.append((1 - time_pred[i])/time_test[i])

    # accuracy
    test_accuracy, idx_correct = accuracy(strategy_pred, strategy_test)

    # Create dataframes to return
    df = pd.DataFrame({'problem': [problem.name],
                       'radius': [problem.radius],
                       'k': [k],
                       'num_var': [num_var],
                       'num_constr': [num_constr],
                       'num_test': [num_test],
                       'num_train': [num_train],
                       'n_theta': [n_theta],
                       'n_corect': [np.sum(idx_correct)],
                       'n_active_sets': [n_active_sets],
                       'accuracy': [accuracy],
                       'n_infeas': [np.sum(infeas >= TOL)],
                       'avg_infeas': [np.mean(infeas)],
                       'avg_subopt': [np.mean(subopt[np.where(infeas <= TOL)])],
                       'max_infeas': [np.max(infeas)],
                       'max_subopt': [np.max(subopt)],
                       'avg_time_improv': [np.mean(time_comp)],
                       'max_time_improv': [np.maximum(time_comp)]
                       })
    df_detail = pd.DataFrame({
        'problem': [problem.name] * num_test,
        'correct': [idx_correct],
        'infeas': infeas,
        'subopt': subopt,
        'time_improvement': time_comp
        })

    return df, df_detail
